src/Database/Database.php

```diff
@@ -564,15 +559,33 @@ public function p(string $sql)
 					break;
 				}
 
-				foreach (array_keys($args) as $param) {
+				// Additional safety check: ensure parameter count matches placeholder count
+				$placeholder_count = substr_count($sql, '?');
+				if (count($args) != $placeholder_count) {
+					// Fall back to direct query with parameter replacement for safety
```

Comment:
When the number of parameters don't match, we possibly should directly raise an error in the log. This is something that needs to be addressed, since this is a crucial bug. Also I don't really like the idea of not using `bindParam` or `bindValue`, since this a safety issue.